### ISSUE TITLE
[request] ability to supply already pre-calculated md5 sum before copy-to-remote

### ISSUE THREAD

**Issue by ribasushi** [rclone/rclone#6480]
#### What is your current rclone version (output from `rclone version`)?
`rclone v1.59.2-DEV` ( result of `go install github.com/rclone/rclone@v1.59.2` ) 

#### What problem are you are trying to solve?
R2 (cloudflare) uploads work great now, thank you for the feature! I am trying to store really large files for which I already have md5 pre-calculated. It would be great to be able to do something like `--checksum=md5deadbeef` to skip the costly re-calculation of something I already have.

#### How do you think rclone should be changed to solve that?
Allow `--checksum` to take a parameter *OR* make a new one `--expect-checksum` or somesuch. The latter would make it more in line with the rest of the UX, e.g.: `rclone cat --expect-checksum=deadcoffee ...`

Thank you!

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
This should be something the [hasher backend](https://rclone.org/hasher/) can take care of.

However I think it may not work at the moment - see this dicussion on the forum: https://forum.rclone.org/t/local-and-hasher-dont-work-nicely-each-other/33179/

@Lesmiscore ?

**Comment by Lesmiscore** [comment]
Hasher may not work if the wrapped backend (in this case R2-I don't know what exactly is) reported MD5 is available

**Comment by ncw** [comment]
R2 runs under S3

**Comment by Lesmiscore** [comment]
https://github.com/rclone/rclone/blob/c5109408c02db3ca14a9a5382373066299c39d25/backend/s3/s3.go#L3817
Indeed it has MD5

**Comment by ncw** [comment]
@ribasushi I understand you have the files on local disk. So the plan would be to use wrap your local backend in a hasher and pre-seed it with a SUM file: https://rclone.org/hasher/#pre-seed-from-a-sum-file

With the modifications to hasher that @Lesmiscore is thinking about, this would then avoid the costly checksum calculation on upload.

How does that sound?

**Comment by ribasushi** [comment]
@ncw @lesmiscore yeah, this will totally work! More specifically: the thing to upload is created by a process which spits out the md5 as it goes. Something like:
```
... someStreamer | zstd -c --long -15 -T0 | tee result.zst | md5sum > result.zst.md5  
```
so at the end of this I have `result.zst` *AND* `result.zst.md5`. When the `.zst` is over a TiB, recomputing the md5 is wasteful at best.

What are next steps here to make this a reality?

**Comment by ncw** [comment]
The next steps would be to see if @Lesmiscore would like to do the changes?

**Comment by Lesmiscore** [comment]
I don't think no changes is needed since OP only should use Hasher with the local directory as its upstream (since they're trying to copy **from local** to R2)
[The issue described here](https://github.com/rclone/rclone/issues/6480#issuecomment-1266719802) is for when there's an Union behind Hasher and local, so most likely it doesn't apply for this issue.

**Comment by ribasushi** [comment]
Ah I see... ok, then I will try to test this some time next week and will report on this ticket. Thanks for the pointers, the Hashers thing is not obviously findable right now
